src/session-state.ts

```diff
@@ -0,0 +1,187 @@
+import { normalizeFlow, sortBreadcrumbs } from "./logic";
+import { ComponentType, SectionStatuses } from "./types";
+import type {
+  NodeId,
+  Breadcrumbs,
+  OrderedBreadcrumbs,
+  NormalizedCrumb,
+  FlowGraph,
+  NormalizedFlow,
+  NormalizedNode,
+  SectionOverview,
+} from "./types";
+
+// TODO
+// SessionState loads a flow and provides an interface for
+// creating, updating and exporting a set of breadcrumbs
+// export class SessionState {}
+
+// StaticSessionState loads a flow and provides an interface for
+// querying the state of a given set of OrderedBreadcrumbs
+export class StaticSessionState {
+  flow: FlowGraph;
+  normalizedFlow: NormalizedFlow;
+
+  constructor(flow: FlowGraph) {
+    this.flow = flow;
+    this.normalizedFlow = normalizeFlow(flow);
+  }
+
+  get sections() {
+    return this.normalizedFlow.filter(
+      (node) => node.type == ComponentType.Section
+    );
+  }
+
+  sortBreadcrumbs(unsortedBreadcrumbs: Breadcrumbs): OrderedBreadcrumbs {
+    return sortBreadcrumbs(this.normalizedFlow, unsortedBreadcrumbs);
+  }
+
+  currentNodes(breadcrumbs: OrderedBreadcrumbs): NormalizedNode[] {
+    if (breadcrumbs.length == 0) {
+      return [];
+    }
+    const currentIndex = this.currentIndex(breadcrumbs);
+    const nextNode = this.normalizedFlow[currentIndex]!;
+    return this.findAllRelatedNodes(nextNode);
+  }
+
+  nextNodes(breadcrumbs: OrderedBreadcrumbs): NormalizedNode[] {
+    let individualNextNode: NormalizedNode | null = null;
+    // get the first node if breadcrumbs are empty
+    if (breadcrumbs.length == 0) {
+      individualNextNode = this.normalizedFlow[0];
+    } else {
+      // otherwise get the first remaining node
+      const remaining = this.remainingNodes(breadcrumbs);
+      individualNextNode = remaining.length ? remaining[0] : null;
+    }
+    if (individualNextNode == null) return [];
+    // return the related nodes
+    return this.findAllRelatedNodes(individualNextNode);
+  }
+
+  findIds(nodeIds: Array<NodeId>): Array<NormalizedNode> {
+    return this.normalizedFlow.filter((node) => nodeIds.includes(node.id));
+  }
+
+  findAllRelatedNodes(node: NormalizedNode): Array<NormalizedNode> {
+    if (node.edges?.length) {
+      return [node, ...this.findIds(node.edges)];
+    }
+    return [node];
+  }
+
+  remainingNodes(breadcrumbs: OrderedBreadcrumbs): Array<NormalizedNode> {
+    if (breadcrumbs.length == 0) {
+      return this.normalizedFlow;
+    }
+    const currentIndex = this.currentIndex(breadcrumbs);
+    const currentNode = this.normalizedFlow[currentIndex]!;
+    return this.normalizedFlow.filter((node, index) => {
+      const afterCurrent = index > currentIndex;
```

Comment:
I may well have missed this somewhere but how are we thinking of handling checklists which make a journey "revisit" another (earlier) forking point?

Reply:
@DafyddLlyr thanks - I'd love to pick this up in more detail on the dev call (if possible). I do have a `previousNodes` method on another branch but I suspect you're talking about a non-obvious behavior in going forwards. Happy to add some more test cases around this as we go to cover these kinds of scenarios.